### DISCUSSION TITLE
`connection.execute` behaves differently depending on whether first line of query is a comment or not

### DISCUSSION THREAD

**Discussion by the21st** [sqlalchemy/sqlalchemy#6706] [Usage Questions] [answered]
**Describe the bug**
An UPDATE query in postgres behaves differently based on whether its first line is a SQL comment.

**To Reproduce**
Provide your [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) example
here.

```py
from sqlalchemy.engine import create_engine

def run_query(query):
    bind_params = {}
    engine = create_engine('postgresql://postgres:password@localhost/postgres')
    connection = engine.connect()

    try:
        connection.execute(query, params=bind_params)
    finally:
        connection.close()
        engine.dispose()

def run_query_and_fetchall(query):
    bind_params = {}
    engine = create_engine('postgresql://postgres:password@localhost/postgres')
    connection = engine.connect()

    try:
        res = connection.execute(query, params=bind_params)
        return res.fetchall()
    finally:
        connection.close()
        engine.dispose()

select_query = """
SELECT content
FROM comments
WHERE id = '20a413be-3b04-4293-9a97-9a12b234660c'
"""

update_query_1 = """
-- this is a comment
UPDATE comments
SET content = 'B'
WHERE id = '20a413be-3b04-4293-9a97-9a12b234660c'
"""

update_query_2 = """
UPDATE comments
SET content = 'C'
WHERE id = '20a413be-3b04-4293-9a97-9a12b234660c'
"""
        
print(run_query_and_fetchall(select_query))
run_query(update_query_1)
print(run_query_and_fetchall(select_query))
run_query(update_query_2)
print(run_query_and_fetchall(select_query))
```

Output:
```
[('A',)]
[('A',)]
[('C',)]
```

Expected output:
either
```
[('A',)]
[('B',)]
[('C',)]
```
or
```
[('A',)]
[('A',)]
[('A',)]
```

**Versions.**
 - OS: MacOS 10.14.6
 - Python: 3.9.2
 - SQLAlchemy: 1.4.20
 - Database: Postgres
 - DBAPI: psycopg2

**Have a nice day!**

**Comment by gordthompson** [comment]
I am able to reproduce your results. However, I get the expected results if I use

```py
def run_query(query):
    engine = create_engine(connection_uri)
    with engine.begin() as connection:
        connection.execute(text(query))
```

**Comment by zzzeek** [comment, marked as answer]
hey there -

you are seeing [library level (emulated) autocommit](https://docs.sqlalchemy.org/en/14/core/connections.html#library-level-e-g-emulated-autocommit) occurring for one of your statements and not the other, as this is based on a regular expression search that does not take things like SQL comments into account.

Solution:  don't use the above feature as it is deprecated for reasons like this and will be removed in SQLAlchemy 2.0.   Use [the 1.4/2.0 API and the newly documented patterns](https://docs.sqlalchemy.org/en/14/tutorial/dbapi_transactions.html) for new code.  for your existing code, change it like this:

```py

engine = create_engine('postgresql://postgres:password@localhost/postgres')

def run_query(query):
    bind_params = {}
    with engine.begin() as conn:
        connection.execute(query, params=bind_params)
   
```

also note above, the engine is created outside the scope of individual functions and is established at module scope.   create_engine() produces a registry that should be long lived within an application.

**Reply by the21st** [reply]
Thanks!
